src/Http/Provider/AbstractHttpProvider.php

```diff
@@ -57,7 +58,32 @@ public function __construct(HttpClient $client, MessageFactory $factory = null)
      */
     protected function getUrlContents(string $url): string
     {
-        $request = $this->getMessageFactory()->createRequest('GET', $url);
+        $request = $this->getRequest($url);
+
+        return $this->getRequestContents($request);
+    }
+
+    /**
+     * @param string $url
+     *
+     * @return RequestInterface
+     */
+    protected function getRequest(string $url): RequestInterface
+    {
+        return $this->getMessageFactory()->createRequest('GET', $url);
+    }
+
+    /**
+     * Send request and return contents. If content is empty, an exception will be thrown.
+     *
+     * @param RequestInterface $request
+     *
+     * @return string
+     *
+     * @throws InvalidServerResponse
+     */
+    protected function getRequestContents(RequestInterface $request): string
+    {
         $response = $this->getHttpClient()->sendRequest($request);
```

Comment:
I don’t think this function should send the request.

Reply:
We could change `getRequestContents(RequestInterface $request)` to `parseResponse(ResponseInterface $response)`, and call `$response = $this->getHttpClient()->sendRequest($request)` inside `getUrlContents` before calling `parseResponse`. The drawback would be that the request url would not be available inside `parseResponse`, so the method would require an extra parameter `$url` in order to throw `InvalidServerResponse`.

Reply:
Oh. I see. Hm. You *could* throw the exception and the catch it...

If you decide to still send the request in this function, I still think you should rename it. Maybe to “getparsedResponse” or something similar. A request does not really have “content”. It has a body of it is a POST request.

Reply:
Good point. I'll rename it. If a provider needs to access the response, it would be possible to move `$response = $this->getHttpClient()->sendRequest($request);` into a separate method `sendRequest`, that could be overridden by the provider.